pkg/lightning/backend/local/local.go

```diff
@@ -1380,25 +1405,95 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID) erro
 		}
 
 		// start to write to kv and ingest
-		err = local.writeAndIngestByRanges(ctx, lf, ranges, remains)
+		err = local.writeAndIngestByRanges(ctx, lf, unfinishedRanges)
 		if err != nil {
 			log.L().Error("write and ingest engine failed", log.ShortError(err))
 			return err
 		}
-
-		unfinishedRanges := remains.take()
-		if len(unfinishedRanges) == 0 {
-			break
-		}
-		log.L().Info("ingest ranges unfinished", zap.Int("remain ranges", len(unfinishedRanges)))
-		ranges = unfinishedRanges
 	}
 
 	log.L().Info("import engine success", zap.Stringer("uuid", engineUUID),
 		zap.Int64("size", lfTotalSize), zap.Int64("kvs", lfLength))
 	return nil
 }
 
+func (engine *File) unfinishedRanges(ranges []Range) []Range {
+	engine.finishedRanges.Lock()
+	defer engine.finishedRanges.Unlock()
+
+	engine.finishedRanges.ranges = dedupAndMergeRanges(engine.finishedRanges.ranges)
+
+	return filterOverlapRange(ranges, engine.finishedRanges.ranges)
+}
+
+func dedupAndMergeRanges(ranges []Range) []Range {
```

Comment:
can it deduplicate, and handle one range totally covers another?

Reply:
Yes，If several range overlap with each other, the result will be the smallest start and the largest end

Reply:
the "dedup" makes me think of "[[1,2], [1,2]] -> [[1,2]]" 😂

Reply:
The `TestDedupAndMergeRanges` unit test has test cases that covers this condition

Reply:
OK, I rename the function name to `sortAndMergeRanges`😅